unmaintained/lunar_calibrate.py

```diff
@@ -0,0 +1,115 @@
+parser.add_option('-g', '--gaincal-duration', type='float', default=60,
+                  help='Minimum duration to track gain calibrator per visit, in seconds '
+                       '(default="%default")')
+parser.add_option('-m', '--max-duration', type='float',
+                  help='Maximum duration of script, in seconds '
+                       '(the default is to keep observing until all sources have set)')
+parser.set_defaults(description='Imaging run', nd_params='off')
+# Parse the command line
+opts, args = parser.parse_args()
+
+# Check options and arguments, and build KAT configuration, connecting to proxies and devices
+if len(args) == 0:
+    raise ValueError("Please specify the target(s) and calibrator(s) to observe as arguments, either as "
+                     "description strings or catalogue filenames")
+
+with verify_and_connect(opts) as kat:
+    sources = collect_targets(kat, args)
+
+    user_logger.info("Imaging targets are [%s]",
+                     ', '.join([repr(target.name) for target in sources.filter(['~bpcal', '~gaincal'])]))
+    user_logger.info("Bandpass calibrators are [%s]",
+                     ', '.join([repr(bpcal.name) for bpcal in sources.filter('bpcal')]))
+    user_logger.info("Gain calibrators are [%s]",
+                     ', '.join([repr(gaincal.name) for gaincal in sources.filter('gaincal')]))
+    duration = {'target': opts.target_duration, 'bpcal': opts.bpcal_duration,
+                'gaincal': opts.gaincal_duration}
+
+    with start_session(kat, **vars(opts)) as session:
+        session.standard_setup(**vars(opts))
+        session.capture_start()
+
+        start_time = time.time()
+        # If bandpass interval is specified, force the first visit to be to the bandpass calibrator(s)
+        time_of_last_bpcal = 0
+        time_of_last_lunar_cal = 0
+        loop = True
+        source_total_duration = defaultdict(float)
+        
+        while loop:
+            # Loop over sources in catalogue in sequence
+            source_observed = defaultdict(bool)
+            for source in sources:
+                # If it is time for a bandpass calibrator to be visited on an interval basis, do so
+                if opts.bpcal_interval is not None and time.time() - time_of_last_bpcal >= opts.bpcal_interval:
+                    time_of_last_bpcal = time.time()
+                    for bpcal in sources.filter('bpcal'):
+                        session.label('track')
+                        track_status = session.track(bpcal, duration=duration['bpcal'])
+                        
+                        if track_status:
+                            source_total_duration[bpcal] += duration['bpcal']
+                if time.time() - time_of_last_lunar_cal >= opts.lunar_interval:
+                    time_of_last_lunar_cal = time.time()
+                    session.label('Moon')
+                    track_status = session.track(katpoint.Target('Moon,special'), duration=duration['bpcal'])
```

Comment:
@spassmoor does this assign a central RA, DEC to the moon? I need to be able to dump out the moon scans with katdal to calibrate it with a combination of CASA and Meqtrees (implementing a disk polarization model).

Reply:
The idea behind my implementation is to do snapshots limiting the movement smearing to a fraction of the synthesized beam

Reply:
The export code in mvftoms.py would need to be modified to either export the Moon as a valid target(quite a lot of work), or generate a target with a ra,dec that is equivalent to the position of the moon( quite easy).

Reply:
In order to predict I need reasonably accurate uv coordinates (up to calculated smearing limits). I thought that having a series of radec targets in the observation itself would avoid modification of mvftoms.

Reply:
The key code to modify would be around line 505 in mvftoms.py 
`                if scan_state != 'track':
                    if options.verbose:
                        print("scan %3d (%4d samples) skipped '%s' - not a track"
                              % (scan_ind, scan_len, scan_state))
                    continue
                if scan_len < 2:
                    if options.verbose:
                        print("scan %3d (%4d samples) skipped - too short"
                              % (scan_ind, scan_len))
                    continue
                if target.body_type != 'radec':
                    if options.verbose:
                        print("scan %3d (%4d samples) skipped - target '%s' not RADEC"
                              % (scan_ind, scan_len, target.name))
                    continue
`
then changing the code to not skip the scan and then changing the target to a fixed radec target it would be fine.

Reply:
Ok I will try implement something like that tomorrow. We will need a switch to dump samples based on total angular movement to make this work correctly.